File: dotnet/src/Rig/Glob.cs
using System.Text.RegularExpressions;

namespace Rig;

/// <summary>
/// Minimal, dependency-free glob matching for config patterns: <c>*</c> matches
/// any run of characters, <c>?</c> a single one. Case-insensitive, anchored
/// (the whole input must match). Used by <c>exclude</c> project filtering.
/// </summary>
internal static class Glob
{
    public static bool IsMatch(string pattern, string input)
    {
        var regex = "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";
        return Regex.IsMatch(input, regex, RegexOptions.IgnoreCase);
    }
}

File: node/src/glob.ts
/**
 * Minimal glob → RegExp for `exclude` matching against package names and
 * relative paths. Supports `*`, `?`, and `**`. Case-insensitive, anchored.
 */
export function globToRegExp(pattern: string): RegExp {
  let re = '^'
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!
    if (ch === '*') {
      if (pattern[i + 1] === '*') {
        re += '.*'
        i++
        if (pattern[i + 1] === '/') i++ // consume the slash after **
      } else {
        re += '[^/]*'
      }
    } else if (ch === '?') {
      re += '[^/]'
    } else if ('.+^${}()|[]\\'.includes(ch)) {
      re += `\\${ch}`
    } else {
      re += ch
    }
  }
  re += '$'
  return new RegExp(re, 'i')
}

/** True if `value` matches the glob `pattern`. */
export function matchGlob(pattern: string, value: string): boolean {
  return globToRegExp(pattern).test(value)
}

/** True if `value` matches any of the patterns. */
export function matchAnyGlob(patterns: string[], value: string): boolean {
  return patterns.some((p) => matchGlob(p, value))
}
